src/algebraic_geometry/EllipticCurve.lean

```diff
@@ -33,55 +33,110 @@ Some more details of the construction can be found on pages 66-69 of
 globally defined by a cubic equation valid over the entire base.
 
 ## TODO
 
-Define the R-points (or even A-points if A is an R-algebra). Care will be needed
-at infinity if R is not a field. Define the group law on the R-points. (hard) prove associativity.
-
+Define the `R`-points (or even `A`-points if `A` is an `R`-algebra). Care will be needed at infinity
+if `R` is not a field. Define the group law on the `R`-points. Prove associativity (hard).
 -/
 
-/-- The discriminant of the plane cubic `Y^2+a1*X*Y+a3*Y=X^3+a2*X^2+a4*X+a6`. If `R` is a field
-then this polynomial vanishes iff the cubic curve cut out by this equation is singular. -/
-def EllipticCurve.disc_aux {R : Type*} [comm_ring R] (a1 a2 a3 a4 a6 : R) : R :=
--432*a6^2 + ((288*a2 + 72*a1^2)*a4 + (-216*a3^2 + (144*a1*a2 + 36*a1^3)*a3 + (-64*a2^3 -
-48*a1^2*a2^2 - 12*a1^4*a2 - a1^6)))*a6 + (-64*a4^3 + (-96*a1*a3 + (16*a2^2 + 8*a1^2*a2 + a1^4))*a4^2
-+ ((72*a2 - 30*a1^2)*a3^2 + (16*a1*a2^2 + 8*a1^3*a2 + a1^5)*a3)*a4 + (-27*a3^4 + (36*a1*a2 +
-a1^3)*a3^3 + (-16*a2^3 - 8*a1^2*a2^2 - a1^4*a2)*a3^2))
+universe u
+
+/-- The discriminant of an elliptic curve given by the long Weierstrass equation
+  `y² + a₁xy + a₃y = x³ + a₂x² + a₄x + a₆`. If `R` is a field, then this polynomial vanishes iff the
+  cubic curve cut out by this equation is singular. Sometimes only defined up to sign in the
+  literature; we choose the sign used by the LMFDB. For more discussion, see
+  [the LMFDB page on discriminants](https://www.lmfdb.org/knowledge/show/ec.discriminant). -/
+def EllipticCurve.disc_aux {R : Type u} [comm_ring R] (a₁ a₂ a₃ a₄ a₆ : R) : R :=
+-(a₁ ^ 2 + 4 * a₂) ^ 2 * (a₁ ^ 2 * a₆ + 4 * a₂ * a₆ - a₁ * a₃ * a₄ + a₂ * a₃ ^ 2 - a₄ ^ 2)
+  - 8 * (2 * a₄ + a₁ * a₃) ^ 3 - 27 * (a₃ ^ 2 + 4 * a₆) ^ 2
+  + 9 * (a₁ ^ 2 + 4 * a₂) * (2 * a₄ + a₁ * a₃) * (a₃ ^ 2 + 4 * a₆)
 
--- If Pic(R)[12]=0 then this definition is mathematically correct
+-- If Pic(R)[12] = 0 then this definition is mathematically correct
 /-- The category of elliptic curves over `R` (note that this definition is only mathematically
-correct for certain rings, for example if `R` is a field or a PID). -/
-structure EllipticCurve (R : Type*) [comm_ring R] :=
-(a1 a2 a3 a4 a6 : R)
-(disc_unit : Rˣ)
-(disc_unit_eq : (disc_unit : R) = EllipticCurve.disc_aux a1 a2 a3 a4 a6)
+  correct for certain rings, for example if `R` is a field or a PID). -/
+structure EllipticCurve (R : Type u) [comm_ring R] :=
+(a₁ a₂ a₃ a₄ a₆ : R) (disc : Rˣ) (disc_eq : ↑disc = EllipticCurve.disc_aux a₁ a₂ a₃ a₄ a₆)
 
 namespace EllipticCurve
 
-instance : inhabited (EllipticCurve ℚ) := ⟨⟨0,0,1,-1,0, ⟨37, 37⁻¹, by norm_num, by norm_num⟩,
-  show (37 : ℚ) = _ + _, by norm_num⟩⟩
+instance : inhabited (EllipticCurve ℚ) :=
+⟨⟨0, 0, 1, -1, 0, ⟨37, 37⁻¹, by norm_num1, by norm_num1⟩, show (37 : ℚ) = _ + _, by norm_num1 ⟩⟩
+
+variables {R : Type u} [comm_ring R] (E : EllipticCurve R)
+
+@[simp] lemma coe_disc : (E.disc : R) = disc_aux E.a₁ E.a₂ E.a₃ E.a₄ E.a₆ := E.disc_eq
```

Comment:
Isn't this lemma identical to `E.disc_eq`? Why not just rename `E.dist_eq` to `E.coe_dist`?

Reply:
This has an extra `@[simp]`?

Reply:
I've scrapped this in favour of writing out `disc_aux` explicitly using the new coefficients.